Update runc to 1.2.0

kolyshkin opened the pull request:
Let's use latest and greatest runc v1.2.0 (both the binary and the set of vendored libraries).

thaJeztah committed:
Dockerfile: update to runc v1.2.0

Release notes:
final: https://github.com/opencontainers/runc/releases/tag/v1.2.0
rc.3:  https://github.com/opencontainers/runc/releases/tag/v1.2.0-rc.3
rc.2:  https://github.com/opencontainers/runc/releases/tag/v1.2.0-rc.2
rc.1:  https://github.com/opencontainers/runc/releases/tag/v1.2.0-rc.1

Breaking changes and deprecations are included below;

Breaking changes:

Several aspects of how mount options work has been adjusted in a way that
could theoretically break users that have very strange mount option strings.
This was necessary to fix glaring issues in how mount options were being
treated. The key changes are:

- Mount options on bind-mounts that clear a mount flag are now always
  applied. Previously, if a user requested a bind-mount with only clearing
  options (such as rw,exec,dev) the options would be ignored and the
  original bind-mount options would be set. Unfortunately this also means
  that container configurations which specified only clearing mount options
  will now actually get what they asked for, which could break existing
  containers (though it seems unlikely that a user who requested a specific
  mount option would consider it "broken" to get the mount options they
  asked foruser who requested a specific mount option would consider it
  "broken" to get the mount options they asked for). This also allows us to
  silently add locked mount flags the user did not explicitly request to be
  cleared in rootless mode, allowing for easier use of bind-mounts for
  rootless containers.
- Container configurations using bind-mounts with superblock mount flags
  (i.e. filesystem-specific mount flags, referred to as "data" in
  mount(2), as opposed to VFS generic mount flags like MS_NODEV) will
  now return an error. This is because superblock mount flags will also
  affect the host mount (as the superblock is shared when bind-mounting),
  which is obviously not acceptable. Previously, these flags were silently
  ignored so this change simply tells users that runc cannot fulfil their
  request rather than just ignoring it.

Deprecated

- runc option --criu is now ignored (with a warning), and the option will
  be removed entirely in a future release. Users who need a non-standard
  criu binary should rely on the standard way of looking up binaries in
  $PATH.
- runc kill option -a is now deprecated. Previously, it had to be specified
  to kill a container (with SIGKILL) which does not have its own private PID
  namespace (so that runc would send SIGKILL to all processes). Now, this is
  done automatically.
- github.com/opencontainers/runc/libcontainer/user is now deprecated, please
  use github.com/moby/sys/user instead. It will be removed in a future
  release.

[@kolyshkin: set CC and CFLAGS, add xx-verify for runc-dmz, use
EXTRA_BUILDTAGS (no need to specify default tags like seccomp)].

Co-authored-by: Kir Kolyshkin <kolyshkin@gmail.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>
Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
An except from s390x failure:
```
2024-07-11T23:39:23.7252075Z  > [runc-build 3/3] RUN --mount=from=runc-src,src=/usr/src/runc,rw     --mount=type=cache,target=/root/.cache/go-build,id=runc-build-linux/s390x <<EOT (set -e...):
2024-07-11T23:39:23.7259197Z 0.307 rm -f libcontainer/dmz/binary/runc-dmz
2024-07-11T23:39:23.7260452Z 0.310 go generate -tags "seccomp urfave_cli_no_docs runc_nodmz" ./libcontainer/dmz
2024-07-11T23:39:23.7263004Z 0.315 go build -trimpath   -tags "seccomp urfave_cli_no_docs runc_nodmz netgo osusergo" -ldflags "-X main.gitCommit=v1.2.0-rc.2-0-gf2d2ee5 -X main.version=1.2.0-rc.2 -extldflags -static " -o runc .
2024-07-11T23:39:23.7264531Z 7.507 # runtime/cgo
2024-07-11T23:39:23.7265202Z 7.507 cc1: error: bad value 'z196' for '-march=' switch
2024-07-11T23:39:23.7281398Z 7.507 cc1: note: valid arguments to '-march=' switch are: nocona core2 nehalem corei7 westmere sandybridge corei7-avx ivybridge core-avx-i haswell core-avx2 broadwell skylake skylake-avx512 cannonlake icelake-client rocketlake icelake-server cascadelake tigerlake cooperlake sapphirerapids alderlake bonnell atom silvermont slm goldmont goldmont-plus tremont knl knm x86-64 x86-64-v2 x86-64-v3 x86-64-v4 eden-x2 nano nano-1000 nano-2000 nano-3000 nano-x2 eden-x4 nano-x4 k8 k8-sse3 opteron opteron-sse3 athlon64 athlon64-sse3 athlon-fx amdfam10 barcelona bdver1 bdver2 bdver3 bdver4 znver1 znver2 znver3 btver1 btver2 native
2024-07-11T23:39:23.7288810Z 39.29 make: *** [Makefile:105: static-bin] Error 1
```

From here, it is clearly seen that gcc for x86_64 is used, instead of the one for the target platform (s390x). The `-march` flag comes from Go build.

kolyshkin commented:
The problem might be in https://github.com/tonistiigi/xx; will keep looking tomorrow. @tonistiigi perhaps can suggest something here?

tonistiigi commented:
https://github.com/opencontainers/runc/blob/3778ae603c706494fd1e2c2faf83b406e38d687d/cc_platform.mk tries to do some of own magic/overrides for cross-compile so not sure what the end result will be.

rata commented:
@tonistiigi thanks! It seems the cc_platform.mk was causing the build issues! 

I'm testing in: https://github.com/moby/moby/pull/48161 with this fix in runc: https://github.com/rata/runc/commits/rata/moby-fixes

kolyshkin commented:
Testing https://github.com/opencontainers/runc/pull/4346 here...

kolyshkin commented:
Hmm, i was not able to reproduce the failures locally. In CI, the following tests fail:
 - TestBuildUserNamespaceValidateCapabilitiesAreV2
 - TestNetworkLocalhostTCPNat
 - TestReadPluginNoRead
 - TestDiff
 
(and maybe some others; I haven't checked all logs).

Locally, these are passing. Maybe I need to set something (running it as `make TEST_SKIP_INTEGRATION_CLI=yes  test-integration` in a Debian Bookworm VM with docker-ce 27.0.3-1~debian.12~bookworm installed.

cyphar commented:
@tonistiigi It only overrides `CC` if `CC` wasn't already set for the purposes of cross-compilation. It seems Docker uses `xx-go --wrap` to hide that it is cross-compiling when building dependencies, which leads to issues. I would expect `CC=xx-clang` (or `CC=clang`, because of `xx-go --wrap`) to fix the issue, though I haven't tested it.

Unfortunately, just removing `cc_platform.mk` will mean that "standard" cross-compilation (`GOARCH=foo make`) on regular systems won't work anymore (to be fair, the script probably only works on openSUSE and Debian).

tonistiigi commented:
@cyphar `xx-go` does not set `CC` for the system (it doesn't require user to set it). It configures `go env CC` instead. And if you run `xx-go --wrap` it sets the go/cgo configuration in the `go env GOENV` where go reads it. `GOARCH` etc is also loaded through `go env`. I think the default for projects should still be to use the configured toolchain of the system. Maybe if you put something like `ifeq ($(origin GOARCH), command line)` in Makefile, then it is clear that the config was passed to the makefile specifically and has the extra behavior of modifying the toolchain.

kolyshkin commented:
The issue seems to be that
1. `xx-go --wrap` writes `~/.config/go/env` file which sets various variables for Go, including `GOARCH`, `CC`, `GOGCCFLAGS` etc. For example, for `s390x` those are:
 - GOARCH='s390x'
 - GCCGO='gccgo'
 - AR='s390x-linux-gnu-ar'
 - CC='s390x-linux-gnu-gcc'
 - PKG_CONFIG='s390x-linux-gnu-pkg-config'
 - GOGCCFLAGS='-fPIC -m64 -march=z196 -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build1204069782=/tmp/go-build -gno-record-gcc-switches'
3. runc's `cc_platform.mk` exports `CC=gcc`. Apparently a local environment has a preference over what's in GOENV file.

Now, gcc tries to use x86_64 gcc with s390x cflags.

tonistiigi commented:
@kolyshkin I don't think xx writes `GOGCCFLAGS` https://github.com/tonistiigi/xx/blob/master/src/xx-go#L104-L128 . The rest of it seems correct .

cyphar commented:
Ideally we would just do `GOARCH=xxx go env CC` to get the compiler to use, but that doesn't work at all outside of `xx-go --wrap`:

 1. For proper target triples, Go doesn't autodetect anything and you need to specify a CC and so on anyway. This does get copied to `go env CC` but that's just passthrough and there's no change in behaviour.
 2. For `GOARCH=386` (and similar `-m` compilation targets) Go doesn't even provide the right CFLAGS for you. They are in `GOGCCFLAGS` but those contain other bits that we don't need.

So we need to parse `GOARCH` for at least case (2). The `GOARCH=... make` thing is nice but it's not really required given that `GOARCH=... go build` doesn't work out of the box AFAICS, so I would accept us breaking this if it meant we could remove.

I suspect Docker doesn't build for `386` but the issue still stands -- we don't want to build runc with an embedded `runc-dmz` that is the wrong architecture. It would be great if there was a supported way to get the compiler options Go wants to use for CGo... We could just use that then.

kolyshkin commented:
Modified this to just set CC and CFLAGS explicitly, this way runc's `cc_platform.mk` doesn't do any guessing, and no changes to runc are needed.

kolyshkin commented:
Full(er) list of test failures:

- TestBuildUserNamespaceValidateCapabilitiesAreV2
- TestNetworkNat
- TestNetworkLocalhostTCPNat
- TestBuildSquashParent
- TestDiff
- TestUsernsCommit
- TestReadPluginNoRead/explicitly_enabled_caching

Most failures are related to not getting output from a container run. As I can't repro this locally I'm not sure how to bisect it.

rata commented:
Same here, if I run things locally (changing the setting so the container uses the iptables uses the nft backend, as that is what my host is using) those tests work fine here too.

However, it is 100% reliable that runc 1.2.x fails here on CI and 1.1.x works just fine.

@thaJeztah do you have any pointers on how to further debug this issue we observe only on CI?

I can try to create a VM with ubuntu, but I hope there is a better way (I don't have any VM setup ready to go now).

kolyshkin commented:
> I can try to create a VM with ubuntu

That's exactly what I was doing (Ubuntu 20.04 in vagrant libvirt) and I still can't repro.

I suspect there are either some limits in CI that we hit, or maybe something like apparmor config.

kolyshkin commented:
The common part between most (or all?) of the failures is not getting the container's output.

Also, I tried again and failed to repro this locally in a Ubuntu 20.04 VM.

AkihiroSuda commented:
> I suspect there are either some limits in CI that we hit, or maybe something like apparmor config.

dmesg may show some hints?
Unloading AppArmor may make some difference?

AkihiroSuda commented:
> any other ideas to try?

Maybe bisect (manually? 😞 )?

AkihiroSuda commented:
Possible reasons:
- ENOSPC
- ENOSPC, but for inodes
- OOM

AkihiroSuda commented:
I'm trying to bisect the commits in https://github.com/moby/moby/pull/48366 but it is quite painful due to compilation failures of dmz and other irrelevant CI failures

kolyshkin commented:
Just found out something I was doing back in 2018 (commit ad2f88d8cc) and decided to try it here.

AkihiroSuda committed:
integration: add wait

Cherry-picked several WIP commits from
https://github.com/moby/moby/commits/b0a592798f4d9d7162f8aedca89ada3a29d60e2c/

Originally-authored-by: Rodrigo Campos <rodrigoca@microsoft.com>
Co-Authored-by: Kir Kolyshkin <kolyshkin@gmail.com>
Signed-off-by: Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>
Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
vendor: bump runc to v1.2.0

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin committed:
daemon: use OwnCgroupPath in withCgroups

Note: this usage comes from commit 56f77d5ade (part of PR 23430).

cgroups.InitCgroupPath is removed from runc (see [1]), and it is
suggested that users use OwnCgroupPath instead, because using init's is
problematic when in host PID namespace (see [2]) and is generally not
the right thing to do (see [3]).

[1]: https://github.com/opencontainers/runc/commit/fd5debf3
[2]: https://github.com/opencontainers/runc/commit/2b28b3c2
[3]: https://github.com/opencontainers/runc/commit/54e20217

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
A failure in `test / test (graphdriver) / unit (firewalld)`:
```
=== Failed
=== FAIL: libnetwork/networkdb TestNetworkDBCRUDTableEntries (7.60s)
    networkdb_test.go:313: Entry existence verification test failed for node2(8ba4809e3fe4)
```

Looks like it's a known issue: https://github.com/moby/moby/pull/47553

rata commented on integration/container/nat_test.go:
```diff
@@ -29,7 +29,14 @@ func TestNetworkNat(t *testing.T) {
 	startServerContainer(ctx, t, msg, 8080)
 
 	endpoint := getExternalAddress(t)
+	try := 0
+retry:
 	conn, err := net.Dial("tcp", net.JoinHostPort(endpoint.String(), "8080"))
+	if err != nil && try < 10 {
+		try++
+		time.Sleep(200 * time.Millisecond)
+		goto retry
+	}
```
I think these changes didn't help. I ended up doing this: https://github.com/rata/moby/commit/2b723e7ff7f915244a8b1295ba5daaeaca20c747

There is a bug saying these tests are flaky (not just the one skipped, IIRC) and I just added the skip like other tests on this file do. You can verify this, but my memory is that at least not all of the changes in this file helped to make the tests green

rata replied:
@kolyshkin sorry, with "these changes" I meant just the retry here in this file. The other changes are still needed, or at least I've seen them flaky with runc 1.2.0 final release in a quick test.

The only difference is that with runc 1.2.0-rcX the tests failed reliably. Now with the overlayfs it is faster and they are just flaky. But I think the rest of the changes to moby tests are needed

rata reviewed (commented):
@kolyshkin thanks for moving this forward!
